tsdb/db_test.go

```diff
@@ -9517,3 +9518,85 @@ func TestBlockClosingBlockedDuringRemoteRead(t *testing.T) {
 	case <-blockClosed:
 	}
 }
+
+var sizeRestrictedFSAvailable = flag.Bool("test.size-restricted-fs", false, "run tests that require size restricted FS")
+
+// TestNoSpaceLeftFailureRecovery uses the size-restricted filesystems in /mnt/test_prom_size_restricted_fs
+// to reproduce an ENOSPC failure during appending, it then frees some space and ensures
+// that appending resumes and that the already produced WAL isn't corrupted.
+// Issues such as https://github.com/prometheus/prometheus/issues/13027 have suggested that the WAL
+// becomes corrupted when the disk is full. This test will help give additional attention to such cases.
+// and help provide an environement to learn more about them.
+func TestSizeRestrictedFS_NoSpaceLeftFailureRecovery(t *testing.T) {
+	fsRoot := "/mnt/test_prom_size_restricted_fs"
+	if !*sizeRestrictedFSAvailable {
+		t.Skip("test requires size restricted FS")
+	}
+
+	for i := range 100 {
+		t.Run(fmt.Sprintf("fs_%d", i), func(t *testing.T) {
+			t.Parallel()
+			root := filepath.Join(fsRoot, fmt.Sprintf("fs_%d", i), "tmp")
+			require.NoError(t, os.RemoveAll(root))
+			require.NoError(t, os.MkdirAll(root, 0o777))
+			ctx := context.Background()
+
+			// Reserve some space in the FS.
+			placeholder := filepath.Join(root, "placeholder")
+			f, err := os.OpenFile(placeholder, os.O_WRONLY|os.O_CREATE, 0o777)
+			require.NoError(t, err)
+			err = fileutil.Preallocate(f, 2*1024*1024, false)
+			require.NoError(t, err)
+			require.NoError(t, f.Close())
+
+			dbDir := filepath.Join(root, t.TempDir())
```

Comment:
[Nit] Isn't it better to use `os.MkdirTemp` here, since the temporary directory will be created beneath `root` (with the guarantee that it's unique)? No big deal, the following just seems safer to me:

```suggestion
			dbDir, err := os.MkdirTemp(root, "")
			require.NoError(t, err)
```

Reply:
sure, I can do that